### PR TITLE
feat: subcommand preconditions

### PR DESCRIPTION
**Pull request by UndiedGamer**

### CODE REVIEW

**Comment by UndiedGamer** [comment]
i need test subjects for this

**Comment by c43721** [comment]
blocked until result v2 refactor

**Comment by c43721** [comment]
@UndiedGamer Please rebase and continue work, this is no longer blocked.

File: packages/subcommands/src/lib/Subcommand.ts [hunk 1]
```diff
@@ -187,6 +201,18 @@ export class Subcommand<PreParseReturn extends Args = Args, O extends Subcommand
 				this.container.client.emit(SubcommandPluginEvents.ChatInputSubcommandRun, interaction, casted, payload);
 				let result: unknown;
 
+				if (casted.preconditions) {
+					const preconditions = new PreconditionContainerArray(casted.preconditions);
+					const result = await preconditions.chatInputRun(interaction, this, context);
+					if (!result.success) {
```
Comment (UndiedGamer) [inline packages/subcommands/src/lib/Subcommand.ts]:
```suggestion
					if (result.isErr()) {
```

File: packages/subcommands/src/lib/Subcommand.ts [hunk 2]
```diff
@@ -156,6 +156,20 @@ export class Subcommand<PreParseReturn extends Args = Args, O extends Subcommand
 				this.container.client.emit(SubcommandPluginEvents.MessageSubcommandRun, message, casted, payload);
 				let result: unknown;
 
+				if (casted.preconditions) {
+					const preconditions = new PreconditionContainerArray(casted.preconditions);
+					const result = await preconditions.messageRun(message, this, context);
+					if (!result.success) {
```
Comment (UndiedGamer) [inline packages/subcommands/src/lib/Subcommand.ts]:
```suggestion
					if (result.isErr()) {
```

**Comment by c43721** [comment]
@UndiedGamer Needs rebase

**Comment by UndiedGamer** [comment]
Moving this to v4 as discussed on discord

**Pull request closed without merging**